File: star-frontend/src/lib/tonalpohualli.js
// Tonalpohualli Calculator for Aztec Day Signs and Galactic Tones
// Based on the traditional 260-day Aztec sacred calendar
import { getDailyToneMessage, getGalacticTone, getToneCompatibility } from './zodiacSystems';

/**
 * Calculate Aztec Day Sign and Sacred Number based on birthdate using the Tonalpohualli calendar.
 * 
 * @param {Date} birthDate - The birth date
 * @returns {Object} Aztec sign information including day sign and sacred number
 */
export function calculateAztecSign(birthDate) {
    // Aztec Day Signs in correct order (20 signs)
    const daySigns = [
        'Crocodile', 'Wind', 'House', 'Lizard', 'Serpent', 'Death', 'Deer',
        'Rabbit', 'Water', 'Dog', 'Monkey', 'Grass', 'Reed', 'Jaguar',
        'Eagle', 'Vulture', 'Movement', 'Flint', 'Storm', 'Sun'
    ];
    
    // Base correlation date (GMT correlation constant)
    // This is based on the most commonly used correlation
    const baseDate = new Date(1970, 0, 1); // Unix epoch as reference
    const baseTonalpohualli = 0; // Adjust based on correlation
    
    // Calculate days since base date
    const daysSinceBase = Math.floor((birthDate - baseDate) / (1000 * 60 * 60 * 24));
    
    // Calculate position in 260-day Tonalpohualli cycle
    const tonalpohalliPosition = (daysSinceBase + baseTonalpohualli) % 260;
    
    // Calculate day sign (0-19) and sacred number (1-13)
    const daySignIndex = tonalpohalliPosition % 20;
    const sacredNumber = (tonalpohalliPosition % 13) + 1;
    
    const daySign = daySigns[daySignIndex];
    
    return {
        daySign: daySign,
        sacredNumber: sacredNumber,
        tonalpohalliPosition: tonalpohalliPosition,
        description: `${sacredNumber} ${daySign}`
    };
}

/**
 * Get the element associated with an Aztec day sign.
 * @param {string} daySign - The Aztec day sign
 * @returns {string} The element
 */
export function getAztecElement(daySign) {
    const elements = {
        'Crocodile': 'Earth', 'Wind': 'Air', 'House': 'Earth', 'Lizard': 'Water',
        'Serpent': 'Fire', 'Death': 'Fire', 'Deer': 'Earth', 'Rabbit': 'Air',
        'Water': 'Water', 'Dog': 'Earth', 'Monkey': 'Air', 'Grass': 'Earth',
        'Reed': 'Water', 'Jaguar': 'Fire', 'Eagle': 'Fire', 'Vulture': 'Air',
        'Movement': 'Water', 'Flint': 'Water', 'Storm': 'Air', 'Sun': 'Fire'
    };
    return elements[daySign] || 'Earth';
}

/**
 * Get the personality traits for an Aztec day sign.
 * @param {string} daySign - The Aztec day sign
 * @returns {string} The traits
 */
export function getAztecTraits(daySign) {
    const traits = {
        'Crocodile': 'Primal, nurturing; creative chaos',
        'Wind': 'Communicative, visionary; restless intellect',
        'House': 'Protective, communal; family-oriented',
        'Lizard': 'Agile, resilient; adaptable survivor',
        'Serpent': 'Transformative, wise; passionate healer',
        'Death': 'Rebirth, transitional; philosophical',
        'Deer': 'Graceful, resilient; humanitarian',
        'Rabbit': 'Harmonious, creative; intuitive artist',
        'Water': 'Emotional, purifying; empathic flow',
        'Dog': 'Loyal, guiding; protective leader',
        'Monkey': 'Playful, inventive; trickster artist',
        'Grass': 'Flexible, grounded; resilient grower',
        'Reed': 'Resilient, visionary; skywalker',
        'Jaguar': 'Powerful, intuitive; warrior shaman',
        'Eagle': 'Visionary, free; soaring achiever',
        'Vulture': 'Wise, liberating; forgiving sage',
        'Movement': 'Dynamic, transformative; change-maker',
        'Flint': 'Obsidian-sharp, decisive; cutter of truth',
        'Storm': 'Revealing, purifying; thunderous renewal',
        'Sun': 'Enlightening, radiant; solar creator'
    };
    return traits[daySign] || 'Mysterious, ancient, powerful';
}

/**
 * Get the quality category for an Aztec day sign.
 * @param {string} daySign - The Aztec day sign
 * @returns {string} The quality
 */
export function getAztecQuality(daySign) {
    const qualities = {
        'Primal': ['Crocodile', 'Wind', 'House', 'Lizard', 'Serpent'],
        'Transformative': ['Death', 'Deer', 'Rabbit', 'Water', 'Dog'],
        'Creative': ['Monkey', 'Grass', 'Reed', 'Jaguar', 'Eagle'],
        'Transcendent': ['Vulture', 'Movement', 'Flint', 'Storm', 'Sun']
    };
    
    for (const [quality, signs] of Object.entries(qualities)) {
        if (signs.includes(daySign)) return quality;
    }
    return 'Primal';
}

/**
 * Get the sacred number meanings in Aztec numerology using Galactic Tones.
 * @param {number} sacredNumber - The sacred number (1-13)
 * @returns {Object} Number meaning and energy
 */
export function getSacredNumberMeaning(sacredNumber) {
    const galacticTone = getGalacticTone(sacredNumber);
    return {
        name: galacticTone.name,
        energy: galacticTone.energy,
        meaning: galacticTone.meaning,
        qualities: galacticTone.qualities,
        challenges: galacticTone.challenges,
        color: galacticTone.color,
        symbol: galacticTone.symbol
    };
}

/**
 * Get enhanced Galactic Tone information for a sacred number.
 * @param {number} toneNumber - The Galactic Tone number (1-13)
 * @returns {Object} Complete Galactic Tone data
 */
export function getGalacticToneInfo(toneNumber) {
    const tone = getGalacticTone(toneNumber);
    return {
        ...tone,
        dailyMessage: getDailyToneMessage(toneNumber),
        toneNumber: toneNumber
    };
}

/**
 * Get complete Aztec sign information for a birth date.
 * @param {Date} birthDate - The birth date
 * @returns {Object} Complete Aztec sign information
 */
export function getCompleteAztecSign(birthDate) {
    const signInfo = calculateAztecSign(birthDate);
    const numberMeaning = getSacredNumberMeaning(signInfo.sacredNumber);
    const galacticToneInfo = getGalacticToneInfo(signInfo.sacredNumber);
    
    return {
        ...signInfo,
        element: getAztecElement(signInfo.daySign),
        traits: getAztecTraits(signInfo.daySign),
        quality: getAztecQuality(signInfo.daySign),
        numberMeaning: numberMeaning,
        galacticTone: galacticToneInfo,
        fullDescription: `${signInfo.description} - ${numberMeaning.name} ${signInfo.daySign}`,
        cosmicSignature: `${galacticToneInfo.symbol} ${numberMeaning.name} ${signInfo.daySign}`,
        toneColor: galacticToneInfo.color,
        avatarEffects: galacticToneInfo.avatarEffect,
        particleEffects: galacticToneInfo.particleEffect
    };
}

/**
 * Get Aztec + Galactic Tone compatibility between two complete sign profiles.
 * @param {Object} profile1 - First complete Aztec sign profile
 * @param {Object} profile2 - Second complete Aztec sign profile
 * @returns {Object} Enhanced compatibility information
 */
export function getEnhancedAztecCompatibility(profile1, profile2) {
    const signCompatibility = getAztecCompatibility(profile1.daySign, profile2.daySign);
    const toneCompatibility = getToneCompatibility(profile1.sacredNumber, profile2.sacredNumber);
    
    // Calculate overall compatibility score
    const compatibilityWeights = {
        'High': 3,
        'Medium': 2,
        'Low': 1
    };
    
    const signScore = compatibilityWeights[signCompatibility.elementCompatibility] + 
                     compatibilityWeights[signCompatibility.qualityCompatibility];
    const toneScore = compatibilityWeights[toneCompatibility.level];
    const totalScore = (signScore + toneScore) / 3;
    
    let overallLevel = 'Low';
    if (totalScore >= 2.5) overallLevel = 'High';
    else if (totalScore >= 1.8) overallLevel = 'Medium';
    
    return {
        daySignCompatibility: signCompatibility,
        galacticToneCompatibility: toneCompatibility,
        overallCompatibility: {
            level: overallLevel,
            score: totalScore,
            description: `${profile1.cosmicSignature} and ${profile2.cosmicSignature} compatibility`,
            strengths: [
                signCompatibility.elementCompatibility === 'High' ? 'Strong elemental harmony' : null,
                signCompatibility.qualityCompatibility === 'High' ? 'Aligned life qualities' : null,
                toneCompatibility.compatible ? 'Synchronized galactic tones' : null
            ].filter(Boolean),
            challenges: [
                signCompatibility.elementCompatibility === 'Low' ? 'Contrasting elemental energies' : null,
                signCompatibility.qualityCompatibility === 'Low' ? 'Different life approaches' : null,
                !toneCompatibility.compatible ? 'Divergent tone energies' : null
            ].filter(Boolean)
        }
    };
}

/**
 * Calculate today's Tonalpohualli energy based on current date.
 * @returns {Object} Today's Aztec cosmic energy
 */
export function getTodaysTonalpohalliEnergy() {
    const today = new Date();
    const todaysSign = getCompleteAztecSign(today);
    
    return {
        ...todaysSign,
        dailyReading: `Today's Energy: ${todaysSign.cosmicSignature}`,
        guidance: todaysSign.galacticTone.dailyMessage,
        elementalInfluence: `${todaysSign.element} element brings ${getElementalGuidance(todaysSign.element)}`,
        qualityInfluence: `${todaysSign.quality} quality encourages ${getQualityGuidance(todaysSign.quality)}`
    };
}

/**
 * Get elemental guidance for daily readings.
 * @param {string} element - The element
 * @returns {string} Guidance message
 */
function getElementalGuidance(element) {
    const guidance = {
        'Fire': 'passion, creativity, and dynamic action',
        'Earth': 'grounding, stability, and practical manifestation',
        'Air': 'communication, mental clarity, and inspired movement',
        'Water': 'emotional flow, intuition, and deep healing'
    };
    return guidance[element] || 'cosmic balance and harmony';
}

/**
 * Get quality guidance for daily readings.
 * @param {string} quality - The quality
 * @returns {string} Guidance message
 */
function getQualityGuidance(quality) {
    const guidance = {
        'Primal': 'foundational beginnings and raw creative power',
        'Transformative': 'deep change and spiritual evolution',
        'Creative': 'artistic expression and innovative solutions',
        'Transcendent': 'higher wisdom and cosmic understanding'
    };
    return guidance[quality] || 'spiritual growth and personal development';
}

/**
 * Get Aztec compatibility between two signs.
 * @param {string} sign1 - First Aztec day sign
 * @param {string} sign2 - Second Aztec day sign
 * @returns {Object} Compatibility information
 */
export function getAztecCompatibility(sign1, sign2) {
    const element1 = getAztecElement(sign1);
    const element2 = getAztecElement(sign2);
    const quality1 = getAztecQuality(sign1);
    const quality2 = getAztecQuality(sign2);
    
    // Element compatibility
    const elementCompatibility = {
        'Fire-Air': 'High', 'Air-Fire': 'High',
        'Fire-Fire': 'High', 'Air-Air': 'High',
        'Water-Earth': 'High', 'Earth-Water': 'High',
        'Water-Water': 'High', 'Earth-Earth': 'High',
        'Fire-Water': 'Medium', 'Water-Fire': 'Medium',
        'Air-Earth': 'Medium', 'Earth-Air': 'Medium',
        'Fire-Earth': 'Low', 'Earth-Fire': 'Low',
        'Air-Water': 'Low', 'Water-Air': 'Low'
    };
    
    // Quality compatibility
    const qualityCompatibility = {
        'Primal-Creative': 'High', 'Creative-Primal': 'High',
        'Transformative-Transcendent': 'High', 'Transcendent-Transformative': 'High',
        'Primal-Transformative': 'Medium', 'Transformative-Primal': 'Medium',
        'Creative-Transcendent': 'Medium', 'Transcendent-Creative': 'Medium',
        'Primal-Transcendent': 'Low', 'Transcendent-Primal': 'Low',
        'Transformative-Creative': 'Low', 'Creative-Transformative': 'Low'
    };
    
    const elemKey = `${element1}-${element2}`;
    const qualKey = `${quality1}-${quality2}`;
    
    return {
        elementCompatibility: elementCompatibility[elemKey] || 'Medium',
        qualityCompatibility: qualityCompatibility[qualKey] || 'Medium',
        signs: [sign1, sign2],
        elements: [element1, element2],
        qualities: [quality1, quality2]
    };
}

File: archive/star-backend-root-20251016/tonalpohualli.py
"""
Tonalpohualli Calculator for Aztec Day Signs and Galactic Tones
Based on the traditional 260-day Aztec sacred calendar
"""

import math
from datetime import date, datetime

# Aztec Day Signs in correct order (20 signs)
AZTEC_ZODIAC = {
    0: 'Crocodile', 1: 'Wind', 2: 'House', 3: 'Lizard', 4: 'Serpent', 
    5: 'Death', 6: 'Deer', 7: 'Rabbit', 8: 'Water', 9: 'Dog', 
    10: 'Monkey', 11: 'Grass', 12: 'Reed', 13: 'Jaguar', 14: 'Eagle',
    15: 'Vulture', 16: 'Movement', 17: 'Flint', 18: 'Storm', 19: 'Sun'
}

# Aztec Day Sign Actions/Traits
AZTEC_ZODIAC_ACTIONS = {
    'Crocodile': 'Primal, nurturing; creative chaos',
    'Wind': 'Communicative, visionary; restless intellect',
    'House': 'Protective, communal; family-oriented',
    'Lizard': 'Agile, resilient; adaptable survivor',
    'Serpent': 'Transformative, wise; passionate healer',
    'Death': 'Rebirth, transitional; philosophical',
    'Deer': 'Graceful, resilient; humanitarian',
    'Rabbit': 'Harmonious, creative; intuitive artist',
    'Water': 'Emotional, purifying; empathic flow',
    'Dog': 'Loyal, guiding; protective leader',
    'Monkey': 'Playful, inventive; trickster artist',
    'Grass': 'Flexible, grounded; resilient grower',
    'Reed': 'Resilient, visionary; skywalker',
    'Jaguar': 'Powerful, intuitive; warrior shaman',
    'Eagle': 'Visionary, free; soaring achiever',
    'Vulture': 'Wise, liberating; forgiving sage',
    'Movement': 'Dynamic, transformative; change-maker',
    'Flint': 'Obsidian-sharp, decisive; cutter of truth',
    'Storm': 'Revealing, purifying; thunderous renewal',
    'Sun': 'Enlightening, radiant; solar creator'
}

# Galactic Tones (Sacred Numbers 1-13)
GALACTIC_TONES = {
    1: {
        'name': 'Magnetic',
        'energy': 'Unity',
        'meaning': 'Attracts, unifies, attracts cooperation',
        'qualities': ['Leadership', 'Purpose', 'Unity'],
        'challenges': ['Over-attachment', 'Rigidity'],
        'color': 'Red',
        'symbol': '•'
    },
    2: {
        'name': 'Lunar',
        'energy': 'Polarity',
        'meaning': 'Challenges, stabilizes, cooperation',
        'qualities': ['Balance', 'Cooperation', 'Stability'],
        'challenges': ['Indecision', 'Over-analysis'],
        'color': 'White',
        'symbol': '••'
    },
    3: {
        'name': 'Electric',
        'energy': 'Service',
        'meaning': 'Bonds, activates, service',
        'qualities': ['Service', 'Activation', 'Bonding'],
        'challenges': ['Impatience', 'Scattered energy'],
        'color': 'Blue',
        'symbol': '•••'
    },
    4: {
        'name': 'Self-Existing',
        'energy': 'Form',
        'meaning': 'Defines, measures, forms',
        'qualities': ['Form', 'Definition', 'Measurement'],
        'challenges': ['Limitation', 'Stubbornness'],
        'color': 'Yellow',
        'symbol': '••••'
    },
    5: {
        'name': 'Overtone',
        'energy': 'Radiance',
        'meaning': 'Commands, empowers, radiance',
        'qualities': ['Command', 'Empowerment', 'Radiance'],
        'challenges': ['Domination', 'Ego'],
        'color': 'Red',
        'symbol': '•••••'
    },
    6: {
        'name': 'Rhythmic',
        'energy': 'Equality',
        'meaning': 'Organizes, balances, equality',  
        'qualities': ['Organization', 'Balance', 'Equality'],
        'challenges': ['Perfectionism', 'Rigidity'],
        'color': 'White',
        'symbol': '••••••'
    },
    7: {
        'name': 'Resonant',
        'energy': 'Attunement',
        'meaning': 'Channels, inspires, attunement',
        'qualities': ['Channeling', 'Inspiration', 'Attunement'],
        'challenges': ['Isolation', 'Over-spirituality'],
        'color': 'Blue',
        'symbol': '•••••••'
    },
    8: {
        'name': 'Galactic',
        'energy': 'Integrity',
        'meaning': 'Harmonizes, models, integrity',
        'qualities': ['Harmony', 'Modeling', 'Integrity'],
        'challenges': ['Perfectionism', 'Judgment'],
        'color': 'Yellow',
        'symbol': '••••••••'
    },
    9: {
        'name': 'Solar',
        'energy': 'Intention',
        'meaning': 'Completes, perfects, intention',
        'qualities': ['Completion', 'Perfection', 'Intention'],
        'challenges': ['Impatience', 'Rushing'],
        'color': 'Red',
        'symbol': '•••••••••'
    },
    10: {
        'name': 'Planetary',
        'energy': 'Manifestation',
        'meaning': 'Produces, manifests, perfects',
        'qualities': ['Production', 'Manifestation', 'Perfection'],
        'challenges': ['Materialism', 'Stubbornness'],
        'color': 'White',
        'symbol': '••••••••••'
    },
    11: {
        'name': 'Spectral',
        'energy': 'Liberation',
        'meaning': 'Dissolves, releases, liberation',
        'qualities': ['Dissolution', 'Release', 'Liberation'],
        'challenges': ['Chaos', 'Destruction'],
        'color': 'Blue',
        'symbol': '•••••••••••'
    },
    12: {
        'name': 'Crystal',
        'energy': 'Cooperation',
        'meaning': 'Dedicates, universalizes, cooperation',
        'qualities': ['Dedication', 'Universalization', 'Cooperation'],
        'challenges': ['Over-giving', 'Martyrdom'],
        'color': 'Yellow',
        'symbol': '••••••••••••'
    },
    13: {
        'name': 'Cosmic',
        'energy': 'Presence',
        'meaning': 'Transcends, endures, presence',
        'qualities': ['Transcendence', 'Endurance', 'Presence'],
        'challenges': ['Aloofness', 'Detachment'],
        'color': 'Red',
        'symbol': '•••••••••••••'
    }
}

def calculate_aztec_day_sign_and_tone(birth_date):
    """
    Calculate Aztec Day Sign and Sacred Number based on birthdate using the Tonalpohualli calendar.
    
    Args:
        birth_date (datetime.date or datetime.datetime): The birth date
        
    Returns:
        dict: Aztec sign information including day sign and sacred number
    """
    if isinstance(birth_date, datetime):
        birth_date = birth_date.date()
    
    # Base correlation date (GMT correlation constant)
    base_date = date(1970, 1, 1)  # Unix epoch as reference
    base_tonalpohualli = 0  # Adjust based on correlation
    
    # Calculate days since base date
    days_since_base = (birth_date - base_date).days
    
    # Calculate position in 260-day Tonalpohualli cycle
    tonalpohualli_position = (days_since_base + base_tonalpohualli) % 260
    
    # Calculate day sign (0-19) and sacred number (1-13)
    day_sign_index = tonalpohualli_position % 20
    sacred_number = (tonalpohualli_position % 13) + 1
    
    day_sign = AZTEC_ZODIAC[day_sign_index]
    
    return {
        'day_sign': day_sign,
        'sacred_number': sacred_number,
        'tonalpohualli_position': tonalpohualli_position,
        'description': f"{sacred_number} {day_sign}",
        'traits': AZTEC_ZODIAC_ACTIONS.get(day_sign, 'Mysterious, ancient, powerful'),
        'galactic_tone': GALACTIC_TONES.get(sacred_number, {})
    }

def get_daily_tonalpohualli():
    """Get today's Tonalpohualli energy"""
    today = date.today()
    return calculate_aztec_day_sign_and_tone(today)

def get_aztec_element(day_sign):
    """Get the element associated with an Aztec day sign."""
    elements = {
        'Crocodile': 'Earth', 'Wind': 'Air', 'House': 'Earth', 'Lizard': 'Water',
        'Serpent': 'Fire', 'Death': 'Fire', 'Deer': 'Earth', 'Rabbit': 'Air',
        'Water': 'Water', 'Dog': 'Earth', 'Monkey': 'Air', 'Grass': 'Earth',
        'Reed': 'Water', 'Jaguar': 'Fire', 'Eagle': 'Fire', 'Vulture': 'Air',
        'Movement': 'Water', 'Flint': 'Water', 'Storm': 'Air', 'Sun': 'Fire'
    }
    return elements.get(day_sign, 'Earth')

def get_aztec_quality(day_sign):
    """Get the quality category for an Aztec day sign."""
    qualities = {
        'Primal': ['Crocodile', 'Wind', 'House', 'Lizard', 'Serpent'],
        'Transformative': ['Death', 'Deer', 'Rabbit', 'Water', 'Dog'],
        'Creative': ['Monkey', 'Grass', 'Reed', 'Jaguar', 'Eagle'],
        'Transcendent': ['Vulture', 'Movement', 'Flint', 'Storm', 'Sun']
    }
    
    for quality, signs in qualities.items():
        if day_sign in signs:
            return quality
    return 'Primal'